### ISSUE TITLE
Linux FastForward only works in full-screen with threaded video enabled

### ISSUE THREAD

**Issue by arifogel** [libretro/RetroArch#13478]
## Description
Fast-forward can be toggled but has no effect while in full-screen mode unless threaded video is enabled. Please note that I have ruled out vsync as a contributor by running with [strangle](https://gitlab.com/torkel104/libstrangle) via `strangle -v0 retroarch` (strangle only works on retroarch `gl` video driver), as well as all combinations of vsync settings in `retroarch` and `NVIDIA X Server Settings` with retroarch video drivers `gl` and `glcore` when run without `strangle`. Note that all other retroarch video drivers crash in my environment on program start.

### Expected behavior

When fastforward is activated, the frame rate should increase whether or not threaded video is enabled.

### Actual behavior
The frame rate does not increase in full screen (includes windowed mode full screen) unless threaded video is enabled.
Note that, threaded video need not be enabled for fastforward to function correctly in windowed (non-full-screen) mode

### Steps to reproduce the bug

1. Open retroarch
2. Ensure that threaded video is disabled
3. Launch any game on any core (same behavior for at least `FCEUmm` and `bsnes-mercury Balanced`)
4. Activate fastforward
5. Observe that the frame rate does not increase.

### Bisect Results
Did not attempt to bisect, because the bug has always been present in my environment.
I can do so if required for debugging.

### Version/Commit
- RetroArch: [v1.9.14/06a2367008]

### Environment information
- OS: Ubuntu 18.04.6 LTS
- Graphics card: NVIDIA GeForce RTX 2060
- Graphics driver: NVIDIA proprietary driver version 460.84

**Comment by arifogel** [comment]
I want to add that using threaded video is not an acceptable workaround for me, because it causes choppy video on all cores in my environment.

**Comment by hizzlekizzle** [comment]
Hmm, I can't reproduce this. I can FF in fullscreen without threaded video just fine. Using Ubuntu, installed via snap, glcore.

I didn't try wayland, though. Are you using wayland or X?

**Comment by arifogel** [comment]
I am using unity with compiz if that matters. I assume that means `X`? I read somewhere that Unity would eventually use Mir, but I don't think that includes Unity version 7.
retroarch is installed via PPA.
```
arifogel@sagan:~$echo $XDG_CURRENT_DESKTOP
Unity:Unity7:ubuntu
```
```
arifogel@sagan:~$compiz --version
Compiz 0.9.13.1
```
```
arifogel@sagan:/etc/apt/sources.list.d$grep retro *
libretro-ubuntu-stable-xenial.list:deb http://ppa.launchpad.net/libretro/stable/ubuntu bionic main
libretro-ubuntu-stable-xenial.list:deb-src http://ppa.launchpad.net/libretro/stable/ubuntu bionic main
```
```
arifogel@sagan:~$dpkg -s retroarch
Package: retroarch
Status: install ok installed
Priority: optional
Section: games
Installed-Size: 16893
Maintainer: Libretro Team <libretro@gmail.com>
Architecture: amd64
Version: 1.9.14-r202112060140+06a2367008-152~ubuntu18.04.1
Provides: retroarch-dbg
Depends: libx11-dev, libxrandr-dev, fonts-dejavu-core, libqt5waylandclient5, libvulkan1, libasound2 (>= 1.0.18), libass9 (>= 1:0.13.6), libavcodec57 (>= 7:3.4.8) | libavcodec-extra57 (>= 7:3.4.8), libavformat57 (>= 7:3.4.8), libavutil55 (>= 7:3.4.8), libc6 (>= 2.15), libcaca0 (>= 0.99.beta17-1), libdbus-1-3 (>= 1.9.14), libdrm2 (>= 2.4.62), libegl1, libflac8 (>= 1.3.0), libfontconfig1 (>= 2.12), libfreetype6 (>= 2.2.1), libgbm1 (>= 8.1~0), libgcc1 (>= 1:3.0), libgl1, libjack-jackd2-0 (>= 1.9.10+20150825) | libjack-0.125, libmbedcrypto1 (>= 2.7), libmbedtls10 (>= 2.0), libmbedx509-0 (>= 2.0), libminiupnpc10 (>= 1.9.20140610), libopenal1 (>= 1.14), libpulse0 (>= 0.99.1), libqt5core5a (>= 5.9.0~beta), libqt5gui5 (>= 5.2.0), libqt5network5 (>= 5.0.2), libqt5widgets5 (>= 5.8.0), libsdl2-2.0-0 (>= 2.0.8), libsixel1, libssl1.1 (>= 1.1.0), libstdc++6 (>= 6), libswresample2 (>= 7:3.4.8), libswscale4 (>= 7:3.4.8), libudev1 (>= 183), libusb-1.0-0 (>= 2:1.0.16), libv4l-0 (>= 0.5.0), libwayland-client0 (>= 1.12), libwayland-cursor0 (>= 1.12), libwayland-egl1 (>= 1.15.0), libx11-6, libx11-xcb1, libxext6, libxinerama1, libxkbcommon0 (>= 0.5.0), libxrandr2 (>= 2:1.2.99.2), libxv1, libxxf86vm1, zlib1g (>= 1:1.1.4)
Conffiles:
 /etc/retroarch.cfg 10c0263e6ff267c8037aec071e73940e
Description: Simple frontend for the libretro library
 RetroArch is an open source, multi-platform frontend for the libretro API. It
 can be used as a modular multi-emulator system, game engine, media player and
 3-D technical demonstration. These features are available through libretro
 cores.
 .
 It provides four built-in graphical user interface flavors: RGUI, XMB, Ozone
 and GLUI.
Homepage: https://www.libretro.com/
```

**Comment by arifogel** [comment]
> Hmm, I can't reproduce this. I can FF in fullscreen without threaded video just fine. Using Ubuntu, installed via snap, glcore.
> 
> I didn't try wayland, though. Are you using wayland or X?

What driver vendor and version are you using for your graphics card?

**Comment by arifogel** [comment]
@hizzlekizzle I am comfortable building from source if necessary to debug. I took a quick look at `retroarch.c` for references to fastforward, but didn't see anything obvious that would be affected by threaded video.

**Comment by arifogel** [comment]
So I got thinking about compiz, and decided to go through the settings. I launched `ccsm` (CompizConfig Settings Manager), unticked "Unredirect Fullscreen Windows" in the `Composite` plugin since it sounded maybe relevant.
The help for that option is:
> Allow drawing of fullscreen windows to not be redirected to offscreen pixmaps

![image](https://user-images.githubusercontent.com/7737874/148886795-9b5bb08e-2838-40a2-8116-1533a301dc3d.png)
After that fastforward is working in fullscreen. However, without "Unredirect Fullscreen WIndows", the video is just as choppy as when threaded video is used.

WIth this is mind, I wonder if there is a common compiz-related reason why enabling threaded video and unchecking that compiz option cause choppy video.

Any insight here?

**Comment by arifogel** [comment]
Another update:
If unredirect is left checked, and I disable "Force Composition Pipeline" in NVIDIA X Server settings, then I can fast forward again, but same choppiness as in the other two cases (threaded video, disable unredirect fullscreen windows):
![image](https://user-images.githubusercontent.com/7737874/148890451-6745ddcc-964a-4cb1-8431-a2cee8754c73.png)

**Comment by arifogel** [comment]
Some more digging:
Can get fastforward working with force composition pipeline enabled in the above screen, as long as "allow flipping" is disabled in OpenGL settings tab. However, same problem: choppy video.
Note that in all other tests above, I had "allow flipping" enabled.
My guess is that the real culprit is "allow flipping", and that nvidia opengl settings are only applied to full screen retroarch when both:
- allow unredirect fullscreen windows in is enabled in compiz
- force composition pipeline is enabled in nvidia settings "X Server Display Configuration" (advanced) is checked

I saw a [reddit post](https://old.reddit.com/r/wine_gaming/comments/iib4dt/games_locked_at_60fps_wine_5836tkgpds/) that suggested that "allow flipping" is related to a frame rate lock.

Since "allow flipping" is required on my setup for me to get smooth video, it doesn't look like there is a good solution. I could try another driver version I suppose, but other than that I don't really have any leads.

**Comment by arifogel** [comment]
So if I disable allow flipping, and force vsync via strangle, then the video is smooth and fast forward does not work.
So now I'm thinking that "allow flipping" implies vsync (even though the driver API indicator doesn't show that vsync is on when only allow flipping is checked but vsync is unchecked).
And the real culprit is the obvious one: vsync

Since I can't allow flipping without forcing vsync, I tried unchecking both, and enabling in-game vsync in retroarch.
This gives me smooth video and working fast forrward (which disabled vsync while toggled).

So the solution is:
- Make sure nothing is forcing vsync
  - for nvidia proprietary drivers, this means you must disable both "allow flipping" and "sync to vblank"
    - If for some reason you have programs that need vsync forced, do it via an application profile rather than globally
  - if using compiz, ensure vsync is unchecked in opengl plugin settings
- Enable vsync only within retroarch if desired
  - Then fastforward is actually allowed to disable vsync, and functions properly.

For optimal performance with compiz and nvidia proprietary drivers:
- make sure allow unredirect fullscreen windows is checked in compiz composite plugin settings
- make sure force composition pipeline is checked in nvidia setttings "X Server Display Configuration" when advanced options are showing

Closing, since solved.
